parser_advanced_test.go

```diff
@@ -0,0 +1,386 @@
+		rhs, err := parseExprPrec(lex, prec.Right)
+		if err != nil {
+			return nil, err
+		}
+		lhs = ExprBinary{lhs, op, rhs}
+	}
+	return lhs, nil
+}
+
+func TestCustomExprParser(t *testing.T) {
+	type Wrapper struct {
+		Expr Expr `@@`
+	}
+
+	exprParser := mustTestParser(t, &Wrapper{}, participle.UseCustom(parseExprAny))
+
+	requireParseExpr := func(src string, expected Wrapper) {
+		t.Helper()
+		var actual Wrapper
+		require.NoError(t, exprParser.ParseString("", src, &actual))
+		require.Equal(t, expected, actual)
+	}
+
+	requireParseExpr(`1`, Wrapper{ExprNumber{1}})
+	requireParseExpr(`1.5`, Wrapper{ExprNumber{1.5}})
+	requireParseExpr(`"a"`, Wrapper{ExprString{"a"}})
+	requireParseExpr(`(1)`, Wrapper{ExprParens{ExprNumber{1}}})
+	requireParseExpr(`1+1`, Wrapper{ExprBinary{ExprNumber{1}, "+", ExprNumber{1}}})
+	requireParseExpr(`1-1`, Wrapper{ExprBinary{ExprNumber{1}, "-", ExprNumber{1}}})
+	requireParseExpr(`1*1`, Wrapper{ExprBinary{ExprNumber{1}, "*", ExprNumber{1}}})
+	requireParseExpr(`1/1`, Wrapper{ExprBinary{ExprNumber{1}, "/", ExprNumber{1}}})
+	requireParseExpr(`1%1`, Wrapper{ExprBinary{ExprNumber{1}, "%", ExprNumber{1}}})
+	requireParseExpr(`a - -b`, Wrapper{ExprBinary{ExprIdent{"a"}, "-", ExprUnary{"-", ExprIdent{"b"}}}})
+
+	requireParseExpr(`a + b - c * d / e % f`, Wrapper{
+		ExprBinary{
+			ExprIdent{"a"}, "+", ExprBinary{
+				ExprIdent{"b"}, "-", ExprBinary{
+					ExprIdent{"c"}, "*", ExprBinary{
+						ExprIdent{"d"}, "/", ExprBinary{
+							ExprIdent{"e"}, "%", ExprIdent{"f"},
+						},
+					},
+				},
+			},
+		},
+	})
+
+	requireParseExpr(`a * b + c * d`, Wrapper{
+		ExprBinary{
+			ExprBinary{ExprIdent{"a"}, "*", ExprIdent{"b"}},
+			"+",
+			ExprBinary{ExprIdent{"c"}, "*", ExprIdent{"d"}},
+		},
+	})
+
+	requireParseExpr(`(a + b) * (c + d)`, Wrapper{
+		ExprBinary{
+			ExprParens{ExprBinary{ExprIdent{"a"}, "+", ExprIdent{"b"}}},
+			"*",
+			ExprParens{ExprBinary{ExprIdent{"c"}, "+", ExprIdent{"d"}}},
+		},
+	})
+
+	require.Equal(t, "Wrapper = Expr .", exprParser.String())
+}
+
+type (
+	MemberString struct {
+		Value string `@String`
+	}
+
+	MemberNumber struct {
+		Value float64 `@Int | @Float`
+	}
+
+	MemberIdent struct {
+		Name string `@Ident`
+	}
+
+	MemberParens struct {
+		Inner ExprPrecAll `"(" @@ ")"`
+	}
+
+	MemberUnary struct {
+		Op   string      `@("-" | "!")`
+		Expr ExprOperand `@@`
+	}
+
+	MemberBinAddSub struct {
+		Head ExprPrec2             `@@`
+		Tail []MemberBinAddSubCont `@@+`
+	}
+
+	MemberBinAddSubCont struct {
+		Op   string    `@("+" | "-")`
+		Expr ExprPrec2 `@@`
+	}
+
+	MemberBinMulDiv struct {
+		Head ExprPrec3             `@@`
+		Tail []MemberBinMulDivCont `@@+`
+	}
+
+	MemberBinMulDivCont struct {
+		Op   string    `@("*" | "/")`
+		Expr ExprPrec3 `@@`
+	}
+
+	MemberBinRem struct {
+		Head ExprOperand        `@@`
+		Tail []MemberBinRemCont `@@+`
+	}
+
+	MemberBinRemCont struct {
+		Op   string      `@"%"`
+		Expr ExprOperand `@@`
+	}
+
+	ExprPrecAll interface{ exprPrecAll() }
```

Comment:
This is super cool, love it!

It would be good to add (or replace the existing) this to the examples folder.

Reply:
I added this as `_examples/expr3` (and the custom parsing as `_examples/expr4`